isarn-sketches-java/src/main/java/org/isarnproject/sketches/java/TDigest.java

```diff
@@ -0,0 +1,761 @@
+        intShuffle(indexes);
+        int sz = cent.length;
+        double[] oldCent = cent;
+        double[] oldMass = mass;
+        cent = new double[sz];
+        mass = new double[sz];
+        reset();
+        for (int j: indexes) updateLogic(oldCent[j], oldMass[j]);
+    }
+
+    /** Reset this t-digest to an empty state */
+    public final void reset() {
+        nclusters = 0;
+        M = 0.0;
+    }
+
+    private final void newCluster(int j, double x, double w) {
+        double[] newCent = cent;
+        double[] newMass = mass;
+        double[] newFtre = ftre;
+        int sz = cent.length;
+        if (nclusters >= sz) {
+            int szinc = (int)Math.ceil(0.1 * (double)sz);
+            sz += szinc;
+            newCent = new double[sz];
+            newMass = new double[sz];
+            newFtre = new double[1 + sz];
+            System.arraycopy(cent, 0, newCent, 0, j);
+            System.arraycopy(mass, 0, newMass, 0, j);
+        }
+        // arraycopy can handle when cent == newCent
+        System.arraycopy(cent, j, newCent, 1 + j, nclusters - j);
+        System.arraycopy(mass, j, newMass, 1 + j, nclusters - j);
+        // do this after copies above
+        newCent[j] = x;
+        newMass[j] = w;
+        nclusters += 1;
+        cent = newCent;
+        mass = newMass;
+        ftre = newFtre;
+        Arrays.fill(ftre, 0, 1 + nclusters, 0.0);
+        for (int k = 0; k < nclusters; ++k) ftInc(k, mass[k]);
+        M += w;
+    }
+
+    private final int closest(double x) {
+        int j = Arrays.binarySearch(cent, 0, nclusters, x);
+        // exact match, return its index:
+        if (j >= 0) return j;
+        // x is not a cluster center, get its insertion index:
+        j = -(j + 1);
+        // x is to left of left-most cluster:
+        if (j == 0) return j;
+        // x is to right of right-most cluster:
+        if (j == nclusters) return j - 1;
+        // x is between two clusters, return index of closest:
+        double dL = x - cent[j - 1];
+        double dR = cent[j] - x;
+        return (dL < dR) ? (j - 1) : j;
+    }
+
+    /** Obtain the number of clusters in this t-digest 
+     * @return the number of clusters in this t-digest
+     */
+    public final int size() {
+        return nclusters;
+    }
+
+    /** Obtain the total mass sampled by this t-digest
+     * @return the total mass
+     */
+    public final double mass() {
+        return M;
+    }
+
+    /** Obtain the compression setting for this t-digest
+     * @return the compression setting
+     */
+    public final double getCompression() {
+        return C;
+    }
+
+    /** Obtain the maximum discrete setting for this t-digest
+     * @return the maximum discrete setting
+     */
+    public final int getMaxDiscrete() {
+        return maxDiscrete;
+    }
+
+    /** Obtain a reference to this t-digest's cluster center array.
+     * NOTE: this array is not safe to modify, and should be used only in "read-only" mode!
```

Comment:
if this needs to be a public method, it should return a copy of the centers

Reply:
could also name it something like `getCentUnsafe` to hammer the point home even more obviously